app/layersproxymodel.cpp

```diff
@@ -153,3 +114,68 @@ QVariant LayersProxyModel::getData( QModelIndex index, int role ) const
 {
   return sourceModel()->data( index, role );
 }
+
+QgsProject *LayersProxyModel::qgsProject() const
+{
+  return mProject;
+}
+
+void LayersProxyModel::setQgsProject( QgsProject *project )
+{
+  if ( mProject != project )
+  {
+    mProject = project;
+    emit qgsProjectChanged();
+  }
+}
+
+LayersProxyModel::LayerModelTypes LayersProxyModel::modelType() const
+{
+  return mModelType;
+}
+
+void LayersProxyModel::setModelType( LayerModelTypes type )
+{
+  if ( mModelType != type )
+  {
+    mModelType = type;
+
+    applyFilterFunction();
+
+    emit modelTypeChanged();
+  }
+}
+
+LayersModel *LayersProxyModel::model() const
+{
+  return mModel;
+}
+
+void LayersProxyModel::setModel( LayersModel *model )
+{
+  if ( mModel != model )
+  {
+    mModel = model;
+    setSourceModel( mModel );
+    emit modelChanged();
+  }
+}
+
+void LayersProxyModel::applyFilterFunction()
```

Comment:
I suggest
```suggestion
void LayersProxyModel::updateFilterFunction()
```

Would be a better name for this purpose